### DISCUSSION TITLE
Help running concurrent async def functions

### DISCUSSION THREAD

**Discussion by mfallon90-old** [cocotb/cocotb#3280] [Q&A] [answered]
Hello everyone,

I am trying to run some tests on an asynchronous fifo. I would like to have two processes operating independently and concurrently, one process which writes to the fifo in the write clk domain and a separate and concurrent process which reads from the fifo in the read domain.

I have written two async def functions (write and read) which work perfectly well independent of each other, but whenever I try to get them to run at the same time, the simulator (icarus) hangs up and I have to manually exit the simulation, not even the timeout triggers.

I admit that my python isn't the greatest and I'm sure the issue lies with a basic misunderstanding on my part, but I've read all the documentation I can find and still can't get it to work. Any help would be greatly appreciated!

I have included the testbench below

```py3

import random
import cocotb
import logging
import queue
from cocotb.clock import Clock
from cocotb.triggers import RisingEdge, ClockCycles, with_timeout, FallingEdge

class TB():
    def __init__(self, dut, wr_prd, rd_prd):
        self.dut = dut
        self.log = logging.getLogger("cocotb.tb")
        self.log.setLevel(logging.DEBUG)

        cocotb.start_soon(Clock(dut.wr_clk, wr_prd, units='ns').start())
        cocotb.start_soon(Clock(dut.rd_clk, rd_prd, units='ns').start())
        self.dut.wr_data.value = 0
        self.dut.wr_vld.value = 0
        self.dut.rd_rdy.value = 0


    async def write(self, sb, num_loop=1, delay=None):
        for _ in range(num_loop):
            
            data = random.randint(0,255)

            self.dut.wr_vld.value = 1
            self.dut.wr_data.value = data

            if (self.dut.wr_rdy == 1):
                await RisingEdge(self.dut.wr_clk)
                self.dut.wr_vld.value = 0
                self.dut.wr_data.value = 0

            else:
                await with_timeout(RisingEdge(self.dut.wr_rdy), 1000, 'ns')
                await RisingEdge(self.dut.wr_clk)
                self.dut.wr_vld.value = 0
                self.dut.wr_data.value = 0

                sb.put(data)

            if (delay != None):
                clk_cycles = random.randint(0,delay)
                await ClockCycles(self.dut.wr_clk, clk_cycles)


    async def read(self, sb, num_loop=1, delay=None):

        for _ in range(num_loop):
            self.dut.rd_rdy.value = 1

            if (self.dut.rd_vld == 1):
                rd_data = self.dut.rd_data
                await RisingEdge(self.dut.rd_clk)
                self.dut.rd_rdy.value = 0

            else:
                await with_timeout(RisingEdge(self.dut.rd_vld), 1000, 'ns')
                rd_data = self.dut.rd_data
                await RisingEdge(self.dut.rd_clk)
                self.dut.rd_rdy.value = 0

            sb_data = sb.get()

            assert rd_data == sb_data

            if (delay != None):
                clk_cycles = random.randint(0,delay)
                await ClockCycles(self.dut.rd_clk, clk_cycles)


async def cycle_rst_n(rst_n, clk):
    rst_n.setimmediatevalue(0)
    await RisingEdge(clk)
    await RisingEdge(clk)
    rst_n.value = 1
    await RisingEdge(clk)
    await RisingEdge(clk)


WR_PRD      = 24
RD_PRD      = 10
NUM_LOOPS   = 20


@cocotb.test()
async def test_async_fifo(dut):
    """Test for asynchronous fifo"""

    # Declare testbench and scoreboard
    tb = TB(dut, WR_PRD, RD_PRD)
    sb = queue.Queue()

    # Reset design
    await cocotb.start(cycle_rst_n(dut.wr_rst_n, dut.wr_clk))
    await cocotb.start_soon(cycle_rst_n(dut.rd_rst_n, dut.rd_clk))

    # Wait some arbitrary time
    await ClockCycles(dut.wr_clk, 100)

    # These are the two functions I would like to have
    # running concurrently!!!!
    await cocotb.start_soon(tb.write(sb,40,20))
    await cocotb.start_soon(tb.read(sb,40,20))

        
    await ClockCycles(dut.wr_clk, 1000)
    dut._log.info('Test done')

```

**Comment by marlonjames** [comment, marked as answer]
You need to start the concurrent tasks, then either wait the correct amount of time in the main function, or wait on the concurrent tasks to finish.

You are close to what you want, but this is the problem:
```py3
    # These are the two functions I would like to have
    # running concurrently!!!!
    await cocotb.start_soon(tb.write(sb,40,20))
    await cocotb.start_soon(tb.read(sb,40,20))
```
`cocotb.start_soon()` will create the task and return. `await cocotb.start_soon()` will create the task, and then wait for the new task to finish before going to the next line.

What you want is to start both, and then wait. You do this by removing the `await` from each:

```py3
    # These are the two functions I would like to have
    # running concurrently!!!!
    cocotb.start_soon(tb.write(sb,40,20))
    cocotb.start_soon(tb.read(sb,40,20))

    # Wait long enough for tasks to finish
    await ClockCycles(dut.wr_clk, 1000)
    dut._log.info('Test done')
```

**Reply by mfallon90-old** [reply]
Thank you for the clarification! 

It turns our there were two problems, one of which  you correctly pointed out regarding the await statement, but when I changed the code according to your suggestion, it would still hang up.

I then began iteratively working to find what exactly was causing the test to break and it ended up being the Queue. Specifically the following code from the read function

```
           sb_data = sb.get()

            assert rd_data == sb_data
```

After I removed that and the await statements, it worked as expected. any ideas why the queue would cause an issue with the async function?

Regards

**Reply by marlonjames** [reply]
I think because you are using the queue from the standard library. It's thread safe, but the cocotb concurrent tasks are run in the same thread.
So `sb.get()` will block everything waiting for an item in the queue, which can't be added because nothing else runs. So you end up hanging forever.

Cocotb does have some asynchronous queues that would work in your testbench: https://docs.cocotb.org/en/stable/library_reference.html#module-cocotb.queue
You would use `await sb.get()`, and the task would stop executing and wait for another task to add an item to the queue, before continuing.

**Comment by mfallon90-old** [comment]
Thanks for the help, got it working with the async queue! much appreciated
